src/test/coinsviewcacheasync_tests.cpp

```diff
@@ -0,0 +1,186 @@
+
+#include <cstdint>
+#include <memory>
+#include <ranges>
+#include <unordered_set>
+
+BOOST_AUTO_TEST_SUITE(coinsviewcacheasync_tests)
+
+struct NoAccessCoinsView : CCoinsView {
+    std::optional<Coin> GetCoin(const COutPoint&) const override { abort(); }
+};
+
+struct CoinsViewCacheAsyncTest : BasicTestingSetup {
+private:
+    std::unique_ptr<CoinsViewCacheAsync> m_async_cache{nullptr};
+    std::unique_ptr<CBlock> m_block{nullptr};
+
+    CBlock CreateBlock(int32_t num_txs) const noexcept
+    {
+        CBlock block;
+        CMutableTransaction coinbase;
+        coinbase.vin.emplace_back();
+        block.vtx.push_back(MakeTransactionRef(coinbase));
+
+        Txid prevhash{Txid::FromUint256(uint256{1})};
+
+        for (const auto i : std::views::iota(1, num_txs)) {
+            CMutableTransaction tx;
+            Txid txid;
+            if (i % 3 == 0) {
+                txid = Txid::FromUint256(uint256(i));
+            } else if (i % 3 == 1) {
+                txid = prevhash;
+            } else {
+                // Test shortid collisions
+                uint256 u{};
+                std::memcpy(u.begin(), prevhash.ToUint256().begin(), 8);
+                txid = Txid::FromUint256(u);
+            }
+            tx.vin.emplace_back(txid, 0);
+            prevhash = tx.GetHash();
+            block.vtx.push_back(MakeTransactionRef(tx));
+        }
+
+        return block;
+    }
+
+public:
+    explicit CoinsViewCacheAsyncTest(const ChainType& chainType = ChainType::MAIN,
+                              const TestOpts& opts = {})
+        : BasicTestingSetup{chainType, opts}
+    {
+        SeedRandomForTest(SeedRand::FIXED_SEED);
+
+        const auto num_txs{100};
+        m_block = std::make_unique<CBlock>(CreateBlock(num_txs));
+    }
+
+    const CBlock& getBlock() const noexcept { return *m_block; }
+};
+
+void PopulateCache(const CBlock& block, CCoinsViewCache& cache, bool spent = false)
+{
+    for (const auto& tx : block.vtx) {
+        for (const auto& in : tx->vin) {
+            auto outpoint{in.prevout};
+            Coin coin{};
+            if (!spent) coin.out.nValue = 1;
+            BOOST_CHECK_EQUAL(coin.IsSpent(), spent);
+            cache.EmplaceCoinInternalDANGER(std::move(outpoint), std::move(coin));
+        }
+    }
+}
+
+void CheckCache(const CBlock& block, const CCoinsViewCache& cache)
+{
+    uint32_t counter{0};
+    std::unordered_set<Txid, SaltedTxidHasher> txids{};
+    txids.reserve(block.vtx.size() - 1);
+
+    for (const auto& tx : block.vtx) {
+        if (tx->IsCoinBase()) {
+            BOOST_CHECK(!cache.GetPossiblySpentCoinFromCache(tx->vin[0].prevout));
+        } else {
+            for (const auto& in : tx->vin) {
+                const auto& outpoint{in.prevout};
+                const auto should_have{!txids.contains(outpoint.hash)};
+                if (should_have) {
+                    cache.AccessCoin(outpoint);
+                    ++counter;
+                }
```

Comment:
`ConnectBlock` calls `AccessCoin` for every input, shouldn't the test do the same?